feat: One recording file per test

martinalbert opened the pull request:
<!--- Tip: You don't have to remove these comments -->

## Description
<!--- Describe your changes in detail -->
Refactors testing approach for saving recordings:
- save recordings next to test files if jest test instance is specified
- adds new parameter for specifying whether `prepare`, `teardown` or `main`
- group recordings to one file under structure `TestRecordings`
  - grouped by key `<profile>/<provider>/<usecase>`
  - and also by hash
- modify `JestGenerator` to expect primitive parameter
- modify `getGenerator` function and renamed it to `parseTestInstance`
  - returns also function to get path to current test file
  - currently only jest supported
- modify `matchTraffic` function to only run `Matcher` and return impact
- modify `TestRecordings` interface to contain also keys for `<profile>/<provider>/<usecase>`
- refactors `recorder` module
  - split code in multiple files and functions

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Station recordings mess

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- If your changes are only to the internals then make sure any function has its attached documentation updated or added (e.g. descriptive name, doc comment, etc.). If your changes are to user-facing APIs or concepts then make sure README.md is updated accordingly (e.g. command help output). -->
- [x] I have updated the documentation accordingly. If you are changing **code related to user secrets** you need to really make sure that [security documentation](SECURITY.md) is correct.
- [x] I have read the **CONTRIBUTION_GUIDE** document.
- [x] I haven't repeated the code. (DRY)
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

martinalbert committed:
feat: create types for new structure of recordings

martinalbert committed:
feat: use one file per test for recordings
- save recordings next to test files if jest test instance is specified
- group recordings to one file under structure TestRecordings
  - grouped by key <profile>/<provider>/<usecase>
  - and also by hash under usecase
- modify JestGenerator to expect primitive parameter
- modify getGenerator function and renamed it to parseTestInstance
  - returns also function to get path to current test file
  - currently only jest supported
- modify matchTraffic function to only run Matcher and return impact
- modify TestRecordings interface to contain also keys for profile/provider/usecase

martinalbert committed:
chore: format

martinalbert committed:
feat: add mechanism for merging new and current recording files

martinalbert committed:
feat: add logic behind env variable for enabling storing new recordings
- this changes approach of matching:
  - if this new ENV variable is not set - testing library behaves as a development tool and each test run with record mode does not match recording, but rewrites old recording right away
  - if this new ENV variable is set - testing library stores any unmatching new traffic into separate file

martinalbert committed:
feat: optimize function updateRecordings and logging

martinalbert committed:
test: update tests regarding parsing of test instance

martinalbert committed:
test: update tests

Jakub-Vacek commented on src/nock/recorder.ts:
```diff
@@ -118,6 +141,87 @@ export async function loadRecording({
   }
 }
 
+type UpdateResult =
```
There is a lot of functions in one file, I would divided it to more files and also I would move parts of function into separate functions.

martinalbert replied:
good point, I tried to [split it to utils](https://github.com/superfaceai/testing/pull/39/commits/24bae166604386430c7fbf80a5c88b1e2b16e53e)

martinalbert replied:
I might split utils more as I'll write tests for it

martinalbert replied:
@Jakub-Vacek I wrote tests and was able to test it without further splitting, but let me know if more changes are needed.

martinalbert committed:
feat(interfaces): remove unnecassary type

martinalbert committed:
fix(recorder): review fixes
- small refactor of recorder module

martinalbert committed:
feat: move processRecordings into optional options

martinalbert committed:
test(recorder): add tests for recorder module
- add tests for functions startRecording and loadRecording

martinalbert committed:
fix(recorder): move determining logic to separate function

martinalbert committed:
chore(recorder): format and move parsing of recordings file

Jakub-Vacek commented on src/superface-test.interfaces.ts:
```diff
@@ -54,6 +55,8 @@ export type ProcessingFunction = (
 ) => Promise<void> | void;
 
 export interface RecordingProcessOptions {
+  prepare?: boolean;
```
nitpick: this could be typed `kind?: 'prepare' | 'teardown'` right? It is not like you can have both true 🤔

martinalbert replied:
Yes I tried it with enum for it and having booleans just felt better for DX 🤷🏻‍♂️

freaz replied:
Kind is better for extension.

![image](https://user-images.githubusercontent.com/959390/198390414-7a6409dc-ff62-4471-90d0-4215db0d9bd2.jpeg)

Jakub-Vacek commented on src/superface-test.ts:
```diff
@@ -94,12 +104,30 @@ export class SuperfaceTest {
     );
     const processRecordings = options?.processRecordings ?? true;
     const inputVariables = searchValues(testCase.input, options?.hideInput);
+    let recordingsType: RecordingType;
```
if we change the `prepare` and `teardown` property this part would be also simplified to something like: `const recordingsType = options?.recordingsType ?? RecordingType.MAIN`

martinalbert committed:
fix(hash): pass state instead of specific property
- to enable retrieving actual test name

martinalbert committed:
fix: review fixes

martinalbert committed:
feat(errors): add new errors for getRecordings function

martinalbert committed:
chore(recorder): remove unnecassary code

martinalbert committed:
test: add and update tests
- add tests for new errors
- add tests for endRecording
- add tests for recorder utils

martinalbert committed:
wip: test: update tests of superface-test client

martinalbert committed:
chore: update changelog

martinalbert committed:
chore: update readme

martinalbert committed:
chore: change version to beta to enable release

Commit:
chore: release v3.0.0-beta.5

martinalbert committed:
chore: update changelog

martinalbert committed:
feat: save recordings under directory composed of profile id

martinalbert committed:
test: update tests according to new changes

Jakub-Vacek commented on src/superface-test.test.ts:
```diff
@@ -35,22 +48,151 @@ jest.mock('./common/format', () => ({
 }));
 
+// jest.mock('./nock/recorder', () => ({
+//   ...jest.requireActual('./nock/recorder'),
```
remove commented code?

Jakub-Vacek commented on src/superface-test.test.ts:
```diff
@@ -60,88 +202,59 @@ describe('SuperfaceTest', () => {
-          },
-        ]);
-        const endRecSpy = jest.spyOn(nock, 'restore');
+      it.skip('writes and restores recordings', async () => {
```
Not sure why we are skipping this?

martinalbert committed:
Merge branch 'dev' into feature/one-recording-file-per-test

martinalbert committed:
wip: test: update tests for client

martinalbert committed:
test: update tests
- add back tests for hashing

martinalbert committed:
fix: remove details of BoundProfileProvider from asserts

Jakub-Vacek commented on src/nock/recorder.test.ts:
```diff
@@ -1,81 +1,644 @@
+jest.mock('nock');
+
+jest.mock('./recorder.utils', () => ({
+  ...jest.requireActual('./recorder.utils'),
```
nitpick: I think you don't need `requireActual` actual as you are importing only Fns that are mocked. Same for other locations.

martinalbert committed:
fix: review fixes
